apm-lambda-extension/extension/apm_server_test.go

```diff
@@ -94,13 +90,204 @@ func TestPostToApmServerDataNotCompressed(t *testing.T) {
 		apmServerUrl: apmServer.URL + "/",
 	}
 
-	err := PostToApmServer(apmServer.Client(), agentData, &config)
+	err := PostToApmServer(apmServer.Client(), agentData, &config, context.Background())
 	assert.Equal(t, nil, err)
 }
 
-func BenchmarkPostToAPM(b *testing.B) {
-	InitLogger()
+func TestGracePeriod(t *testing.T) {
+	apmServerTransportState.ReconnectionCount = 0
+	val0 := computeGracePeriod().Seconds()
+	assert.Equal(t, val0, float64(0))
+
+	apmServerTransportState.ReconnectionCount = 1
+	val1 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val1, float64(1), 0.1*1)
+
+	apmServerTransportState.ReconnectionCount = 2
+	val2 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val2, float64(4), 0.1*4)
+
+	apmServerTransportState.ReconnectionCount = 3
+	val3 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val3, float64(9), 0.1*9)
+
+	apmServerTransportState.ReconnectionCount = 4
+	val4 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val4, float64(16), 0.1*16)
+
+	apmServerTransportState.ReconnectionCount = 5
+	val5 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val5, float64(25), 0.1*25)
+
+	apmServerTransportState.ReconnectionCount = 6
+	val6 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val6, float64(36), 0.1*36)
+
+	apmServerTransportState.ReconnectionCount = 7
+	val7 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val7, float64(36), 0.1*36)
+}
+
+func TestEnterBackoffFromHealthy(t *testing.T) {
+	SetApmServerTransportStatus(Healthy, 0)
+	// Compress the data
+	pr, pw := io.Pipe()
+	gw, _ := gzip.NewWriterLevel(pw, gzip.BestSpeed)
+	go func() {
+		gw.Write([]byte(""))
+		gw.Close()
+		pw.Close()
+	}()
+
+	// Create AgentData struct with compressed data
+	data, _ := ioutil.ReadAll(pr)
+	agentData := AgentData{Data: data, ContentEncoding: "gzip"}
+
+	// Create apm server and handler
+	apmServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		bytes, _ := ioutil.ReadAll(r.Body)
+		assert.Equal(t, string(data), string(bytes))
+		assert.Equal(t, "gzip", r.Header.Get("Content-Encoding"))
+		w.Write([]byte(`{"foo": "bar"}`))
+	}))
+	apmServer.Close() // Close the APM server early so that POST requests fail and that backoff is enabled
+
+	config := extensionConfig{
+		apmServerUrl: apmServer.URL + "/",
+	}
+
+	PostToApmServer(apmServer.Client(), agentData, &config, context.Background())
+	// No way to know for sure if failing or pending (0 sec grace period)
+	assert.True(t, apmServerTransportState.Status != Healthy)
+	assert.Equal(t, apmServerTransportState.ReconnectionCount, 0)
+}
+
+func TestEnterBackoffFromFailing(t *testing.T) {
+	SetApmServerTransportStatus(Healthy, 0)
```

Comment:
In previous iterations, the tests were failing because the grace period goroutine initialized during the previous test was racing against the `SetApmServerTransportStatus(Healthy, 0)` : we would call the function and set the status of the APM server transport to `Healthy`, only for the goroutine to complete its execution and to change the status of the APM server transport to `Pending`, which is not expected at the start of this test.